### DISCUSSION TITLE
ESP32 based filament scale

### DISCUSSION THREAD

**Discussion by mplogas** [Donkie/Spoolman#127] [Ideas]
I'm currently designing and building a esp32 based filament scale using a load cell. Currently, my solution has a display and an MQTT connection for Home Assistant integration. I have planned adding RFID reader, but my AliExpress delivery is taking a few more days. Also, the 3D models still need some refinement but once I'm happy with the result they will be released among the source code here on GH.

I have found this project by chance yesterday evening browsing the web and think it making it work with spoolman is rather easy and the perfect addition. Before I start the implementing your API, I have a few questions (due to the limited HW of the ESP32 I likely need to drop MQTT in favor of HTTP): 
1. Is MQTT support on your roadmap? Or will it always be HTTP/REST? 
2. Spoolman already has a QR feature, so we're technically half-way to tags. Would this be a valid feature request?
3. The schema would need an extension to incorporate the RFID tag GUID. Do you agree or would you rather suggest reusing on of the existing fields? 

Let me know what you think! I was turned down by the author of 3D-MES yesterday when I asked if he's fine to let me add my solution to his filament management system - hopefully you're not doing the same :)

**Comment by TexZeTech** [comment]
> author of 3D-MES 

This took WAY to long to find that https://www.3d-mes.com their site & discord seems to be German here is a auto translated google link https://www-3d--mes-com.translate.goog/?_x_tr_sl=de&_x_tr_tl=en&_x_tr_hl=en-US&_x_tr_pto=wapp&_x_tr_hist=true

> I'm currently designing and building a esp32 based filament scale using a load cell

I think thats a good idea admittedly I would rather it be integrated with the printer as there is an interface in mainsail and I think fludd, but having options is nice so my two cents is to go for it.

**Comment by mplogas** [comment]
Yes, 3D-MES is the program, their maker has unfortunately turned me down when I offered integrating my DIY scale into his system.  
In the meantime I have been busy adding more features to my smart scale. Feel free to check it out here (https://github.com/mplogas/SmartMass).
Spoolman integration via REST is on my feature list. I'm currently controlling the scale through MQTT (configuration, taring, calibration, tag writing) so I have to figure out if it makes sense to get these features into Spoolman . Until then, Home Assistant (and its UI) could fill the gap.

**Comment by PeteUgrin** [comment]
ESP32 solution is obvious. But RPi in the printer also has GPIO pins, so it is possible to connect the scale directly to RPi. I can technically imagine even multiple scales. But my IT skills concerning Klipper are equal to zero.

**Comment by jebril76** [comment]
Did something like this for octoprint. But it's now discontinued.
https://www.youtube.com/watch?v=lZznkP-F4qI

**Reply by mplogas** [reply]
Heh, neat! Any advice you want to share (like dont's & do's)? I'm iterating (and dog-fooding) since late September and learned a lot, but I'm always looking to learn more - feel free to start a discussion in one of my SmartMass repos if you think this doesn't belong into the SpoolMan discussion.

**Comment by mplogas** [comment]
Thanks for the feedback, everyone!

The project has moved forward and so far I have built the embedded firmware, an MQTT-REST bridge, a.k.a SmartMass controller, a flexible dev drop-in for SpoolMan, and the printable model for the scale. I'm currently in the process of fine-tuning the model and ironing out bugs in both the embedded firmware and MQTT bridge. 
I'll get back with another update once I reach v1.0 (this will include SpoolMan support)

**Comment by Superbrain8** [comment]
ive got something similar in the works. but as a more Standalone approach to allow the use of any firmware, not just supported ones.
Plan is to call the API of Spoolman to update the useage.

**Reply by mplogas** [reply]
I do have a standalone tool for now, called [smartmass controller](https://github.com/mplogas/SmartMass-Controller). i've been busy with work/other projects lately, and smartmass is at a state where it's usable. my spoolman integration is not finished yet, but feel free to either provide PRs to smartmass or use my code for inspiration.

**Reply by joramk** [reply]
Did something similar with an ESP32, a serial connected scale. a HD44780 display and a wireless bluetooth QR code scanner to track filament usage after printing on my Bambu Lab printer.

**Comment by rainman19121979** [comment]
I built this and adapted the original code slightly to my needs. https://github.com/rainman19121979/Spoolman-Scale
